enhancements/api-review/new-CRD-ManagedClusterVersion.md

```diff
@@ -0,0 +1,452 @@
+   **owner** of the `foo` cluster.
+1. The **owner** of the `foo` cluster uses the information presented by **OCM**
+   to make an informed decision about the desired upgrade path for the `foo`
+   cluster, and requests **OCM** to orchestrate that upgrade (via existing means,
+   outside the scope of this document).
+1. When the upgrade is complete, `foo`'s **cluster-version-operator** updates
+   `foo`'s ClusterVersion object with new version and upgrade information.
+1. Hive's **clusterversion controller**, in the course of reconciling the
+   ClusterDeployment object for `foo`, retrieves the ClusterVersion object and,
+   seeing its Status differs from that of the (now extant) ManagedClusterVersion
+   object on the hub, updates the latter.
+1. Later, when the **owner** of the `foo` cluster wishes to perform another
+   upgrade, **OCM** shows an updated view of the available upgrades.
+1. Repeat. Profit.
+
+### API Extensions
+
+```go
+import (
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+// ManagedClusterVersion is the Schema for the managedclusterversions API
+// +k8s:openapi-gen=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=managedclusterversions,shortName=mcv,scope=Namespaced
+type ManagedClusterVersion struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// NOTE: No Spec
+
+	Status configv1.ClusterVersionStatus `json:"status,omitempty"`
+}
+```
+
+Extensions to hive/hypershift/etc and ACM/OCM/etc are outside the scope of this document.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+At the time of this writing, hypershift
+[brokers ClusterVersion Status piecemeal](https://github.com/openshift/hypershift/blob/9adaf7778dd3595ab57e06e4eee7dc3bff353b6b/api/hypershift/v1beta1/hostedcluster_types.go#L2101),
+cloning top-level fields and importing lower-level fields, selectively copying in the pieces
+it deems important. As a result of this enhancement, it is expected that hypershift will
+pivot to a design similar to that described [above](#workflow-description).
+
+Importantly, having a unified solution for brokering this information from the hive/hypershift
+layer up to the \*CM layer is intended to simplify the processing needed by the latter. With
+hive and hypershift both presenting this unified view, \*CM can use common code to retrieve
+and process that object.
+
+#### Standalone Clusters
+
+Standalone clusters are not expected to use this object or carry any instances of it.
+
+#### Single-node Deployments or MicroShift
+
+These are affected only if they themselves are running a fleet manager.
+In that case, each instance of ManagedClusterVersion is simply a CR to
+manage like any other.
+
+If SNO/MicroShift clusters are part of a fleet, their fleet manager may
+broker their ClusterVersion objects in the manner described [above](#workflow-description).
+In this scenario they are the same as any other OpenShift spoke.
```

Comment:
The intent is to use ACM to manage some aspects of MicroShift deployments. I don't know if hive is involved in the integration between ACM and MicroShift.

MicroShift does not have a ClusterVersion API because upgrades are not driven by the CVO. MicroShift uses a ConfigMap to report its version data.

If hive is part of the integration of ACM and MicroShift, will hive have a separate implementation of where to get the version details for MicroShift?

If hive is not present, would something else need to create the ManagedClusterVersion CR in the ACM hub cluster? What will do that?

Reply:
> MicroShift does not have a ClusterVersion API

I didn't realize.

> MicroShift uses a ConfigMap to report its version data.

Does that ConfigMap have the same scope of information as ClusterVersion? Not that we need to get into it here, but if so... why wouldn't we be using CVO?

> If hive is part of the integration of ACM and MicroShift

It's not. Uh, unless Assisted supports MicroShift? Does it?

> If hive is not present, would something else need to create the ManagedClusterVersion CR in the ACM hub cluster? What will do that?

Yes, exactly the point of making this CRD common rather than scoped to hive. In the case of hypershift, the idea is for hypershift to do it. If there are other fleet manager thingies in the world, they would (or could) do the same.

In the ACM scenario, both hive and hypershift would be present, each managing their own subset of clusters, each generating ManagedClusterVersion CRs for their subset, resulting in (identically-schemaed) objects for every spoke the ACM instance manages. That's the dream :)

Reply:
I [confirmed](https://redhat-internal.slack.com/archives/CUPJTHQ5P/p1709766106650909) that Assisted doesn't do MicroShift, so I think we're in the clear here in terms of hive (not) having to understand the ConfigMap thing.

That doesn't mean *ACM* doesn't/won't support MicroShift, but since upgrades there are such a different beast, I don't imagine they'll be using this mechanism at all. I'll update accordingly.